File: lib/ids.js
'use strict';

const PROJECTS = require('../data/projects').map(p => ({ ...p, prefix: p.prefix.toLowerCase() }));
const TD = PROJECTS.find(p => p.prefix === 't_');
const FLAG_FEMALE = 1;
const KNOWN_IDS = {};
const VANILLA_FACTIONS = [
	'Ashlanders', 'Blades', 'Camonna Tong', 'Census and Excise', 'Clan Aundae', 'Clan Berne',
	'Clan Quarra', 'Fighters Guild', 'Hlaalu', 'Imperial Cult', 'Imperial Knights', 'Imperial Legion',
	'Mages Guild', 'Morag Tong', 'Nerevarine', 'Redoran', 'Sixth House', 'Talos Cult', 'Telvanni',
	'Temple', 'Thieves Guild', 'Twin Lamps', 'Dark Brotherhood', 'Hands of Almalexia', 'Royal Guard',
	'East Empire Company', 'Skaal'
];

function checkId(record, mode, id) {
	const project = PROJECTS.find(p => id.startsWith(p.prefix));
	if(!project) {
		console.log(record.type, record.id, 'does not match a known ID scheme');
	} else if(project === TD && mode !== 'TD') {
		console.log(record.type, record.id, 'has a', TD.name, 'ID');
	}
}

module.exports = {
	onRecord(record, mode, recordId) {
		if(mode === 'TD' && record.type === 'Faction' && VANILLA_FACTIONS.includes(record.id)) {
			return;
		}
		if(record.type === 'Bodypart') {
			if(record.data?.vampire && record.data.part === 'Head') {
				const id = `b_v_${record.name}_${record.data.female & FLAG_FEMALE ? 'f' : 'm'}_head_01`;
				if(record.id !== id) {
					console.log(record.type, record.id, 'should have id', id);
				}
			} else {
				checkId(record, mode, recordId);
			}
		} else if(record.id && !['Cell', 'Dialogue', 'LandscapeTexture', 'Region', 'StartScript'].includes(record.type)) {
			checkId(record, mode, recordId);
		}
		if(record.type !== 'Cell' && recordId) {
			if(recordId in KNOWN_IDS) {
				console.log(record.type, record.id, 'shares its ID with a record of type', KNOWN_IDS[recordId].type);
			} else {
				KNOWN_IDS[recordId] = record;
			}
		}
	}
};

File: src/validators/ids.rs
use std::collections::HashMap;

use super::Context;
use crate::{context::Mode, handlers::Handler, util::is_correct_vampire_head};
use tes3::esp::{Bodypart, BodypartFlags, BodypartId, EditorId, TES3Object, TypeInfo};

const VANILLA_FACTIONS: [&str; 27] = [
    "Ashlanders",
    "Blades",
    "Camonna Tong",
    "Census and Excise",
    "Clan Aundae",
    "Clan Berne",
    "Clan Quarra",
    "Fighters Guild",
    "Hlaalu",
    "Imperial Cult",
    "Imperial Knights",
    "Imperial Legion",
    "Mages Guild",
    "Morag Tong",
    "Nerevarine",
    "Redoran",
    "Sixth House",
    "Talos Cult",
    "Telvanni",
    "Temple",
    "Thieves Guild",
    "Twin Lamps",
    "Dark Brotherhood",
    "Hands of Almalexia",
    "Royal Guard",
    "East Empire Company",
    "Skaal",
];

fn is_female(part: &Bodypart) -> bool {
    part.data.flags.contains(BodypartFlags::FEMALE)
}

fn is_vampire_head(part: &Bodypart) -> bool {
    part.data.vampire && (part.data.part == BodypartId::Head)
}

fn check_id(context: &Context, record: &TES3Object) {
    let id = record.editor_id();
    let matching = context.projects.iter().find(|p| p.matches(&id));
    match matching {
        Some(project) => {
            if context.mode != Mode::TD && project.prefix == "T_" {
                println!("{} {} has a {} ID", record.type_name(), id, project.name);
            }
        }
        None => {
            println!(
                "{} {} does not match a known ID scheme",
                record.type_name(),
                id
            );
        }
    }
}

pub struct IdValidator {
    known: HashMap<String, &'static str>,
}

impl Handler<'_> for IdValidator {
    fn on_record(&mut self, context: &Context, record: &TES3Object) {
        match record {
            TES3Object::Bodypart(part) => {
                if is_vampire_head(part) {
                    if !is_correct_vampire_head(&part.id, &part.race, is_female(part)) {
                        println!(
                            "Bodypart {} should have id b_v_{}_{}_head_01",
                            part.id,
                            part.race,
                            if is_female(part) { "f" } else { "m" }
                        );
                    }
                } else {
                    check_id(context, record);
                }
                self.check_known(record);
            }
            TES3Object::Cell(_) => {}
            TES3Object::Dialogue(_) => {
                self.check_known(record);
            }
            TES3Object::Faction(_) => {
                let id: &str = &record.editor_id();
                if context.mode != Mode::TD || !VANILLA_FACTIONS.contains(&id) {
                    check_id(context, record);
                    self.check_known(record);
                }
            }
            TES3Object::GameSetting(_) => {
                println!("Found dirty {} {}", record.type_name(), record.editor_id());
            }
            TES3Object::DialogueInfo(_) => {}
            TES3Object::PathGrid(_) => {}
            TES3Object::Region(_) => {
                self.check_known(record);
            }
            TES3Object::SoundGen(_) => {}
            TES3Object::StartScript(_) => {}
            TES3Object::MagicEffect(mgef) => {
                println!("Found dirty {} {:?}", record.type_name(), mgef.effect_id);
            }
            _ => {
                check_id(context, record);
                self.check_known(record);
            }
        }
    }
}

impl IdValidator {
    pub fn new() -> Self {
        Self {
            known: HashMap::new(),
        }
    }

    fn check_known(&mut self, record: &TES3Object) {
        let typename = record.type_name();
        if let Some(prev) = self
            .known
            .insert(record.editor_id().to_ascii_lowercase(), typename)
        {
            println!(
                "{} {} shares its ID with a record of type {}",
                typename,
                record.editor_id(),
                prev
            );
        }
    }
}
